include/exadg/incompressible_navier_stokes/time_integration/time_int_bdf_pressure_correction.cpp

```diff
@@ -340,83 +340,90 @@ TimeIntBDFPressureCorrection<dim, Number>::momentum_step()
-  }
-
-  /*
-   *  explicit variable viscosity update executed prior to calculation of rhs_momentum
-   */
-  if(this->param.viscous_problem() and this->param.viscosity_is_variable() and
-     this->param.treatment_of_variable_viscosity == TreatmentOfVariableViscosity::Explicit)
-  {
-    dealii::Timer timer_viscosity_update;
-    timer_viscosity_update.restart();
 
-    pde_operator->update_viscosity(velocity_np);
+    bool const update_preconditioner =
+      this->param.update_preconditioner_momentum and
+      ((this->time_step_number - 1) % this->param.update_preconditioner_momentum_every_time_steps ==
+       0);
 
-    if(this->print_solver_info() and not(this->is_test))
+    if(this->param.nonlinear_problem_has_to_be_solved())
     {
-      this->pcout << std::endl << "Update of variable viscosity:";
-      print_wall_time(this->pcout, timer_viscosity_update.wall_time());
+      /*
+       *  Calculate the vector that is constant when solving the nonlinear momentum equation
+       *  (where constant means that the vector does not change from one Newton iteration
+       *  to the next, i.e., it does not depend on the current solution of the nonlinear solver)
+       */
+      VectorType rhs(velocity_np);
+      rhs_momentum(rhs);
+
+      // solve non-linear system of equations
+      auto const iter = pde_operator->solve_nonlinear_momentum_equation(
+        velocity_np,
+        rhs,
+        this->get_next_time(),
+        update_preconditioner,
+        this->get_scaling_factor_time_derivative_term());
+
+      iterations_momentum.first += 1;
+      std::get<0>(iterations_momentum.second) += std::get<0>(iter);
+      std::get<1>(iterations_momentum.second) += std::get<1>(iter);
+
+      if(this->print_solver_info() and not(this->is_test))
+      {
+        this->pcout << std::endl << "Solve momentum step:";
+        print_solver_info_nonlinear(this->pcout,
+                                    std::get<0>(iter),
+                                    std::get<1>(iter),
+                                    timer.wall_time());
+      }
     }
-  }
+    else // linear problem
+    {
+      AssertThrow(this->param.viscous_problem(), dealii::ExcMessage("Should not arrive here."));
 
-  /*
-   *  Calculate the right-hand side of the linear system of equations
-   *  or the vector that is constant when solving the nonlinear momentum equation
-   *  (where constant means that the vector does not change from one Newton iteration
-   *  to the next, i.e., it does not depend on the current solution of the nonlinear solver)
-   */
-  VectorType rhs(velocity_np);
-  rhs_momentum(rhs);
+      /*
+       *  update variable viscosity prior to calculation of rhs_momentum
+       */
+      if(this->param.viscosity_is_variable())
+      {
+        AssertThrow(this->param.treatment_of_variable_viscosity ==
+                      TreatmentOfVariableViscosity::Explicit,
+                    dealii::ExcMessage("Should not arrive here."));
```

Comment:
With reference to comment https://github.com/exadg/exadg/pull/671#discussion_r1646197944, this implementation should now more clearly reflect the underlying logic of the code.